pkg/webhook/server/server.go

```diff
@@ -0,0 +1,295 @@
+	}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/validate", s.handle(s.validate))
+	mux.HandleFunc("/mutate", s.handle(s.mutate))
+	listenerChan := s.startServer(l, internalStopCh, mux)
+
+	if certSourceChan == nil {
+		certSourceChan = blockingChan(internalStopCh)
+	}
+	if healthzChan == nil {
+		healthzChan = blockingChan(internalStopCh)
+	}
+
+	select {
+	case err = <-healthzChan:
+	case err = <-certSourceChan:
+	case err = <-listenerChan:
+	case <-stopCh:
+	}
+
+	close(internalStopCh)
+	shutdown = true
+
+	maybeLog(s.Log).Info("waiting for server to shutdown")
+	waitForAll(healthzChan, certSourceChan, listenerChan)
+	maybeLog(s.Log).Info("server shutdown successfully")
+
+	return err
+}
+
+func (s *Server) startServer(l net.Listener, stopCh <-chan struct{}, handle http.Handler) <-chan error {
+	ch := make(chan error)
+	go func() {
+		defer close(ch)
+
+		srv := &http.Server{
+			Handler: handle,
+		}
+		select {
+		case err := <-channelWrapper(func() error { return srv.Serve(l) }):
+			ch <- err
+		case <-stopCh:
+			// allow a fixed 5s for graceful shutdown
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+			defer cancel()
+			if err := srv.Shutdown(ctx); err != nil {
+				maybeLog(s.Log).Error(err, "failed to gracefully shutdown http server")
+				ch <- err
+			}
+			maybeLog(s.Log).Info("shutdown HTTP server gracefully")
+		}
+	}()
+	return ch
+}
+
+func (s *Server) startCertificateSource(stopCh <-chan struct{}) <-chan error {
+	fn := func() error {
+		return s.CertificateSource.Run(stopCh)
+	}
+	return channelWrapper(fn)
+}
+
+func waitForAll(chs ...<-chan error) error {
+	for _, ch := range chs {
+		if err := <-ch; err != nil {
+			return fmt.Errorf("error waiting for goroutine to exit: %w", err)
+		}
+	}
+	return nil
+}
+
+func channelWrapper(fn func() error) <-chan error {
+	ch := make(chan error)
+	go func() {
+		defer close(ch)
+		ch <- fn()
+	}()
+	return ch
+}
+
+// blockingChan returns a 'no-op' error channel.
+// When stopCh is closed, the error channel will also be closed.
+func blockingChan(stopCh <-chan struct{}) <-chan error {
+	ch := make(chan error)
+	go func() {
+		defer close(ch)
+		<-stopCh
+	}()
+	return ch
+}
+
+func (s *Server) scheme() *runtime.Scheme {
+	if s.Scheme == nil {
+		return defaultScheme
+	}
+	return s.Scheme
+}
+
+func (s *Server) validate(obj runtime.Object) runtime.Object {
+	review := obj.(*admissionv1beta1.AdmissionReview)
+	resp := s.ValidationWebhook.Validate(review.Request)
+	review.Response = resp
+	return review
+}
+
+func (s *Server) mutate(obj runtime.Object) runtime.Object {
+	review := obj.(*admissionv1beta1.AdmissionReview)
+	resp := s.MutationWebhook.Admit(review.Request)
+	review.Response = resp
+	return review
+}
+
+func (s *Server) handle(inner func(runtime.Object) runtime.Object) func(w http.ResponseWriter, req *http.Request) {
+	return func(w http.ResponseWriter, req *http.Request) {
+		defer req.Body.Close()
+
```

Comment:
If we error here, are we writing anything to the response to reflect this?

Reply:
Nope, and you've just reminded me what the big TODO in this code is 🙃

Reply:
done